blog/2020-02-28-db2-cdc-approaches.adoc

```diff
@@ -0,0 +1,248 @@
+purposes.  As it works with a wide range of relational database system and
+is open source we could imagine that database administrators would allow
+it  to  be  used  to  generate a  representation  of  their  data  for
+downstream  applications.   Essentially,  the  Debezium  system  would
+become an  extension of  the database source  system. Debezium  is not
+required to produce an *identical*  copy of the database  tables (unlike
+IIDR or SQL Replication). Typically the downstream application are for
+auxiliary tasks, i.e.  analytics, not for fail  over, meaning problems
+such as preserving precise types are less pressing.  For example, if
+a time-stamp  field is represented  as a  string in ElasticSearch it
+is not the end of the world.
+
+The only concern we had  with Debezium was  that it didn't  have a
+connector for Db2.
+
+Two approaches  presented themselves: 
+
+- Use  the low level Db2  API to read directly the transaction log like  IIDR does.
+- Use the SQL Replication CDC capture tables to read capture tables using SQL.
+
+
+An investigation  of the code concluded that  the model used  by the
+already  existing  connector  for  Microsoft  SQL  Server  could
+be largely reused for Db2. In essence: 
+
+- The SQL queries to poll the changes are different
+- The structure and nature of  the LSN are  different
+- The fact  that Db2 distinguishes between a  database system and  a database  while SQL Server needs to be accounted for.
+
+Otherwise everything else could be reused.
+
+== Future Work/Extensions
+
+=== Benchmarking
+The Db2 Connector and SQL Server use a polling model i.e. the connectors
+periodically query the CDC table to determine what has changed since
+the last time they polled. This is in contrast to MySQL which has a notification model.
+
+We are interested in building a general purpose framework for benchmarking
+systems in order to get a better understanding of where the trade-offs
+are in terms of latency, throughput of the CDC system and load on the
+source system.
+
+
+=== Db2 Notification System
+
+Another approach is to build a notification system on Db2. This would involve:
+
+- Identifying change events by the usage of OS file system watchers (Linux or Windows).
+This can monitor the transaction log directory of the Db2 database and send events when files are modified or created.
+
+- Determining the exact nature of the event by reading the actual table changes with the link:https://www.ibm.com/support/knowledgecenter/en/SSEPGG_11.1.0/com.ibm.db2.luw.apdv.api.doc/doc/r0001673.html[db2ReadLog API]. In principle
+this API can be invoked remotely as a service. 
+
+- Determining the related Db2 data structure via SQL connection, e.g. table DDL.
+
+The debezium event-driven Db2 connector would wait on notifications and then read the actual changes via db2ReadLog and SQL.
+This would require the watcher agent to run locally on the database system, similarly to the capture server.
```

Comment:
Could you add some kind of summary or wrap-up towards the end, perhaps linking back to your initial discussion of your datalake use case? E.g. are you using this in prod already, or are planning to do so etc.

Reply:
Done